docs/root/configuration/other_features/internal_listener.rst

```diff
@@ -0,0 +1,100 @@
+.. _config_internal_listener:
+
+Internal Listener
+=================
+
+How it works
+------------
+
+This extension contains 2 major components to add a listener with
+an :ref:`envoy internal address <envoy_v3_api_msg_config.core.v3.EnvoyInternalAddress>`
+and to create a client connection to that :ref:`listener <envoy_v3_api_msg_config.listener.v3.Listener>`
+
+**envoy.bootstrap.internal_listener_registry** This bootstrap extension is required to support looking up the target listener via an
+:ref:`envoy internal address <envoy_v3_api_msg_config.core.v3.EnvoyInternalAddress>` on each worker threads.
+
+**network.connection.client.envoy_internal** It is a client connection factory. The factory is implicitly instantiated by the dispatcher to establish a client connection to an
```

Comment:
@lambdai what confuses me is that you explain here there are separate components, but there is in fact only a single extension.

The bundling is a bit strange, in that we are mixing upstream and downstream concerns, we typically have had independent registration in situations like this before. On the downside, splitting causes potential for misconfiguration. OTOH it's kind of confusing to have something that is nominally a "bootstrap extension" providing an "internal listener registry" also magically creating a new client connection factory type. This is really just applying principle-of-least-surprise.

Reply:
This client connection factory is useless without this bootstrap extension because there is nowhere to lookup the internal listeners.

The very initial version of this PR is https://github.com/envoyproxy/envoy/pull/18105, which doesn't put this extension functionality into the dispatcher interface. [It was considered as a pollution to the Envoy core.](https://github.com/envoyproxy/envoy/pull/18105#discussion_r766098084) 

Now the bootstrap extension is added to allow Envoy user to disable this functionality of internal listener registration.

The client connection factory is actually the interface to the internal listener. Without it, the extension functionality is not accessible. 

Do you think we should rename this extension bootstrap.internal_client_connection_factory and doesn't change the implementation that provides the per worker thread registry?

Reply:
Maybe some of my reticence can be explained by naming, I agree. It would be fine to have a single `bootstrap.internal_listener` extension that registers both. I just think it strange a "registry" is somehow also magically inserting a client connection factory.

Reply:
Alternatively, what do you think we should do?

Split the two and throw exception when any of them cannot find each other? The client factory doesn't have the luxury to reject LDS/CDS config. 

Ideally we can reject EDS or CDS config that containing internal address, but there is also opinion that rejecting after  valiate each endpoint is expensive. Also the internal address could be referred from any config so it's not realistic to reject a http filter config that contains an internal address at configuration time.

Reply:
re: renaming

make sense to me. Let me update